pkg/cluster/audiences.go

```diff
@@ -0,0 +1,57 @@
+package cluster
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/go-logr/logr"
+	authentication "k8s.io/api/authentication/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	// Default value of audiences for DSCI.SM.auth.
+	defaultAudiences = []string{"https://kubernetes.default.svc"}
+)
+
+func isDefaultAudiences(specAudiences *[]string) bool {
+	return specAudiences == nil || reflect.DeepEqual(*specAudiences, defaultAudiences)
+}
+
+type TokenSupplier func() (string, error)
+
+// GetEffectiveClusterAudiences returns the audiences defined for the cluster
+// falling back to the default audiences in case of errors.
+func GetEffectiveClusterAudiences(cli client.Client, log logr.Logger, specAudiences *[]string, getTokenFunc TokenSupplier) []string {
+	if isDefaultAudiences(specAudiences) {
+		return fetchClusterAudiences(cli, log, getTokenFunc)
+	}
+	return *specAudiences
+}
+
+func fetchClusterAudiences(cli client.Client, log logr.Logger, getTokenFunc TokenSupplier) []string {
+	token, err := getTokenFunc()
+	if err != nil {
+		log.Error(err, "Error getting token, using default audiences")
+		return defaultAudiences
+	}
+
+	tokenReview := &authentication.TokenReview{
+		Spec: authentication.TokenReviewSpec{
+			Token: token,
+		},
+	}
+
+	if err = cli.Create(context.Background(), tokenReview, &client.CreateOptions{}); err != nil {
```

Comment:
Correct, originally we were using opaque token for the bash "workaround" and this solution is based on this, so perhaps, using SA (jwt) token aud is simpler and more efficient. We could replace this impl and use https://github.com/golang-jwt/jwt instead to fetch standard claims without verification (so passing `nil` as the key func ). @israel-hdez @aslakknutsen do you think it's better to rely on the fact that SA token is JWT?

I just checked in ROSA and we can also parse JWT instead. For opaque (openshift oauth) tokens it's the same audience. We can get it by using the exact call to TokenReview API.

This behavior is only invoked when you do not specify anything in DSCI.ServiceMesh.Auth part. If you later do, it will update underlying config map which is now used by `odh-model-controller`